kernel/src/main/scala/cats/kernel/Hash.scala

```diff
@@ -0,0 +1,52 @@
+package cats.kernel
+
+import scala.{specialized => sp}
+import scala.util.hashing.Hashing
+
+/**
+ * A type class used to represent a hashing scheme for objects of a given type.
+ * For any two instances `x` and `y` that are considered equivalent under the
+ * equivalence relation defined by this object, `hash(x)` should equal `hash(y)`.
+ * @author Tongfei Chen
+ */
+trait Hash[@sp A] extends Any with Eq[A] with Serializable { self =>
+
+  /**
+   * Returns the hash code of the given object under this hashing scheme.
+   */
+  def hash(x: A): Int
```

Comment:
The `Object#hashCode: Int` in Java is basically used for distribution within `Set` or `Map`-like data structures and for that purpose `Int` will remain sufficient for some time, for one because the size of an array in Java can't exceed `Int.MaxValue` either and if you instantiate an `Array<Int>` of size `Int.MaxValue` that's around ~ 8 GB of RAM.

So I'm not seeing the size of an array evolving to a `Long`, for one because of backwards compatibility, but also it's probably counter productive to allow contiguous memory blocks bigger than that. For now at least.
I'm guessing that's the reason other languages use 32 bit integers (or sometimes less) for their hash codes as well. For example: https://hackage.haskell.org/package/hashable-1.2.6.1/docs/Data-Hashable.html

The questions are:

1. besides doing distribution in data structures like `HashMap` / `HashSet`, are there any other use cases?
2. doesn't `Long` have performance implications? it's an important question to answer, as HashMaps tend to be abused
3. are there any languages using `Long` for their universal `hashCode` already and if so, I'd be interested in why?